File: kakao.app/js/ked/edit/mode/brckts.js
var _k_ = {eql: function (a,b,s) { var i, k, v; s = (s != null ? s : []); if (Object.is(a,b)) { return true }; if (typeof(a) !== typeof(b)) { return false }; if (!(Array.isArray(a)) && !(typeof(a) === 'object')) { return false }; if (Array.isArray(a)) { if (a.length !== b.length) { return false }; var list = _k_.list(a); for (i = 0; i < list.length; i++) { v = list[i]; s.push(i); if (!_k_.eql(v,b[i],s)) { s.splice(0,s.length); return false }; if (_k_.empty(s)) { return false }; s.pop() } } else if (_k_.isStr(a)) { return a === b } else { if (!_k_.eql(Object.keys(a),Object.keys(b))) { return false }; for (k in a) { v = a[k]; s.push(k); if (!_k_.eql(v,b[k],s)) { s.splice(0,s.length); return false }; if (_k_.empty(s)) { return false }; s.pop() } }; return true }, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, isStr: function (o) {return typeof o === 'string' || o instanceof String}}

var brckts

import kxk from "../../../kxk.js"
let kseg = kxk.kseg
let kutil = kxk.kutil

import belt from "../tool/belt.js"

import mode from "../mode.js"


brckts = (function ()
{
    brckts["autoStart"] = true
    brckts["surround"] = {'#':['#{','}'],'{':['{','}'],'}':['{','}'],'[':['[',']'],']':['[',']'],'(':['(',')'],')':['(',')'],"'":["'","'"],'"':['"','"']}
    function brckts (state)
    {
        this.state = state
    
        this["handleKey"] = this["handleKey"].bind(this)
        this["conditionalSwapStringDelimiters"] = this["conditionalSwapStringDelimiters"].bind(this)
        this["swapStringDelimiters"] = this["swapStringDelimiters"].bind(this)
        this.name = 'brckts'
    }

    brckts.prototype["cursorsSet"] = function ()
    {
        var cursors, lines

        if (this.state.s.selections.length)
        {
            if (_k_.eql(this.allSpans, this.state.s.highlights))
            {
                this.state.setHighlights([])
            }
            return
        }
        if (this.state.s.highlights.length)
        {
            if (_k_.eql(!this.allSpans, this.state.s.highlights.asMutable()))
            {
                return
            }
        }
        lines = this.state.s.lines
        cursors = this.state.s.cursors
        this.openCloseSpans = belt.openCloseSpansForPositions(lines,cursors)
        this.stringDelimiterSpans = belt.stringDelimiterSpansForPositions(lines,cursors)
        this.state.setHighlights(this.openCloseSpans.concat(this.stringDelimiterSpans))
        return this.allSpans = this.state.s.highlights.asMutable()
    }

    brckts.prototype["swapStringDelimiters"] = function ()
    {
        var cursors, selections

        this.state.begin()
        this.state.pushState()
        cursors = this.state.allCursors()
        selections = this.state.allSelections()
        this.state.setSelections(belt.rangesForSpans(this.stringDelimiterSpans))
        this.state.moveCursorsToEndOfSelections()
        if (this.state.textOfSelection().startsWith('"'))
        {
            this.state.insert("'")
        }
        else
        {
            this.state.insert('"')
        }
        this.state.setCursors(cursors)
        this.state.setSelections(selections)
        return this.state.end()
    }

    brckts.prototype["conditionalSwapStringDelimiters"] = function ()
    {
        if (!this.stringDelimiterSpans)
        {
            return
        }
        if (belt.textForSpans(this.state.s.lines,this.stringDelimiterSpans).startsWith("'"))
        {
            return this.swapStringDelimiters()
        }
    }

    brckts.prototype["handleKey"] = function (key, event)
    {
        var delrngs, nsegl, nsegs, pairs, pos, rngs, seg

        switch (key)
        {
            case "alt+cmd+'":
                if (!_k_.empty(this.stringDelimiterSpans))
                {
                    this.swapStringDelimiters()
                    return
                }
                break
            case 'alt+cmd+b':
                if (!_k_.empty(this.stringDelimiterSpans))
                {
                    this.state.setSelections(belt.rangesForSpans(this.stringDelimiterSpans))
                    this.state.moveCursorsToEndOfSelections()
                    return
                }
                if (!_k_.empty(this.openCloseSpans))
                {
                    this.state.setSelections(belt.rangesForSpans(this.openCloseSpans))
                    this.state.moveCursorsToEndOfSelections()
                    return
                }
                break
            case 'delete':
                if (_k_.empty(this.state.s.selections))
                {
                    pairs = kutil.uniq(Object.values(brckts.surround))
                    if (!_k_.empty((rngs = belt.rangesOfPairsSurroundingPositions(this.state.s.lines,pairs,this.state.s.cursors))))
                    {
                        this.state.setSelections(rngs)
                        this.state.deleteSelection()
                        return
                    }
                }
                break
        }

        if (_k_.empty(brckts.surround[event.char]))
        {
            return 'unhandled'
        }
        if (event.char === '#')
        {
            if (!_k_.empty(belt.positionsAndRangesOutsideStrings(this.state.s.lines,this.state.s.selections,this.state.s.cursors)))
            {
                return 'unhandled'
            }
            this.conditionalSwapStringDelimiters()
            this.state.begin()
            if (!_k_.empty(this.state.s.selections))
            {
                this.state.surroundSelection(event.char,brckts.surround[event.char])
            }
            else
            {
                this.state.insert(brckts.surround[event.char][0] + brckts.surround[event.char][1])
                this.state.moveCursors('left')
            }
            this.state.end()
            return
        }
        else
        {
            if (!_k_.empty(this.state.s.selections))
            {
                if (!_k_.empty(this.stringDelimiterSpans))
                {
                    delrngs = belt.normalizeRanges(belt.rangesForSpans(this.stringDelimiterSpans))
                    if (_k_.eql(delrngs, this.state.s.selections))
                    {
                        return 'unhandled'
                    }
                }
                return this.state.surroundSelection(event.char,brckts.surround[event.char])
            }
            nsegl = belt.segsForPositions(this.state.s.lines,this.state.s.cursors)
            nsegs = kutil.uniq(nsegl)
            var list = _k_.list(nsegs)
            for (var _a_ = 0; _a_ < list.length; _a_++)
            {
                seg = list[_a_]
                if (brckts.surround[event.char][1] === event.char)
                {
                    if (seg === event.char)
                    {
                        this.state.moveCursors('right')
                        return
                    }
                }
                if (!(_k_.in(seg,['',undefined,' ','}',']',')'])))
                {
                    return "unhandled"
                }
            }
            var list1 = _k_.list(this.state.s.cursors)
            for (var _b_ = 0; _b_ < list1.length; _b_++)
            {
                pos = list1[_b_]
                if (belt.isUnbalancedPosition(this.state.s.lines,pos,event.char))
                {
                    return "unhandled"
                }
            }
            this.state.insert(brckts.surround[event.char][0] + brckts.surround[event.char][1])
            this.state.moveCursors('left')
            return
        }
        return 'unhandled'
    }

    return brckts
})()

export default brckts;

File: kakao.app/love/lua/edit/mode/brckts.lua
--[[
    ███████    ████████    ███████  ███   ███  █████████   ███████
    ███   ███  ███   ███  ███       ███  ███      ███     ███     
    ███████    ███████    ███       ███████       ███     ███████ 
    ███   ███  ███   ███  ███       ███  ███      ███          ███
    ███████    ███   ███   ███████  ███   ███     ███     ███████ 
--]]


local brckts = class("brckts")
    brckts.static.autoStart = true
    brckts.static.surround = {
        ['#'] = array('#{', '}'), -- <- this has to come, 
        ['{'] = array('{', '}'), -- <- before that (does it?), 
        ['}'] = array('{', '}'), 
        ['['] = array('[', ']'), 
        [']'] = array('[', ']'), 
        ['('] = array('(', ')'), 
        [')'] = array('(', ')'), 
        ["'"] = array("'", "'"), 
        ['"'] = array('"', '"'), 
        -- '<': ['<' '>'] # html , 
        -- '>': ['<' '>'] # html , 
        -- '*': ['*' '*'] # md   
        }


function brckts:init(state) 
        self.state = state
        self.name = 'brckts'
        return self
    end

-- ███   ███  ███   ███████   ███   ███  ███      ███   ███████   ███   ███  █████████
-- ███   ███  ███  ███        ███   ███  ███      ███  ███        ███   ███     ███   
-- █████████  ███  ███  ████  █████████  ███      ███  ███  ████  █████████     ███   
-- ███   ███  ███  ███   ███  ███   ███  ███      ███  ███   ███  ███   ███     ███   
-- ███   ███  ███   ███████   ███   ███  ███████  ███   ███████   ███   ███     ███   


function brckts:cursorsSet() 
        if #self.state.s.selections then 
            if (self.allSpans == self.state.s.highlights) then 
                self.state:setHighlights(array())
            end
            
            return
        end
        
        if #self.state.s.highlights then 
            if (not self.allSpans == self.state.s.highlights) then return end -- .asMutable() # from the previous brckts highlight
        end
        
        local spans, openClose, strings = belt.spansOfNestedPairsAtPositions(self.state.s.lines, self.state.s.cursors)
        
        if empty(spans) then 
            self.openCloseSpans = belt.openCloseSpansForPositions(self.state.s.lines, self.state.s.cursors)
            self.stringDelimiterSpans = belt.stringDelimiterSpansForPositions(self.state.s.cursors)
            spans = (self.openCloseSpans + self.stringDelimiterSpans)
        else 
            self.openCloseSpans = openClose
            self.stringDelimiterSpans = strings
        end
        
        self.state:setHighlights(spans)
        
        self.allSpans = self.state.s.highlights return -- .asMutable()
    end

--  ███████  ███   ███   ███████   ████████ 
-- ███       ███ █ ███  ███   ███  ███   ███
-- ███████   █████████  █████████  ████████ 
--      ███  ███   ███  ███   ███  ███      
-- ███████   ██     ██  ███   ███  ███      


function brckts:swapStringDelimiters() 
        self.state:begin()
        self.state:pushState()
        local cursors = self.state:allCursors()
        local selections = self.state:allSelections()
        self.state:setSelections(belt.rangesForSpans(self.stringDelimiterSpans))
        self.state:moveCursorsToEndOfSelections()
        if kstr.startsWith(self.state:textOfSelection(), '"') then 
            self.state:insert("'")
        else 
            self.state:insert('"')
        end
        
        self.state:setCursors(cursors)
        self.state:setSelections(selections)
        return self.state:ende()
    end


function brckts:conditionalSwapStringDelimiters() 
        if not self.stringDelimiterSpans then return end
        if kstr.startsWith(belt.textForSpans(self.state.s.lines, self.stringDelimiterSpans), "'") then 
            return self:swapStringDelimiters()
        end
    end

-- ███   ███  ████████  ███   ███        
-- ███  ███   ███        ███ ███         
-- ███████    ███████     █████          
-- ███  ███   ███          ███           
-- ███   ███  ████████     ███           


function brckts:handleKey(key, event) 
        if (key == "alt+cmd+'") then 
                if valid(self.stringDelimiterSpans) then 
                    self:swapStringDelimiters()
                    return
                end
        elseif (key == 'alt+cmd+b') then 
                if valid(self.stringDelimiterSpans) then 
                    self.state:setSelections(belt.rangesForSpans(self.stringDelimiterSpans))
                    self.state:moveCursorsToEndOfSelections()
                    return
                end
                
                if valid(self.openCloseSpans) then 
                    self.state:setSelections(belt.rangesForSpans(self.openCloseSpans))
                    self.state:moveCursorsToEndOfSelections()
                    return
                end
        elseif (key == 'delete') then 
                if empty(self.state.s.selections) then 
                    local pairs = util.uniq(dict.values(brckts.surround))
                    local rngs = belt.rangesOfPairsSurroundingPositions(self.state.s.lines, pairs, self.state.s.cursors)
                    if valid(rngs) then 
                        self.state:setSelections(rngs)
                        self.state:deleteSelection()
                        return
                    end
                end
        end
        
        if empty(brckts.surround[event.char]) then return 'unhandled' end
        
        if (event.char == '#') then 
            if valid(belt.positionsAndRangesOutsideStrings(self.state.s.lines, self.state.s.selections, self.state.s.cursors)) then return 'unhandled' end
            
            self:conditionalSwapStringDelimiters()
            
            self.state:begin()
            
            if valid(self.state.s.selections) then 
                self.state:surroundSelection(event.char, brckts.surround[event.char])
            else 
                self.state:insert(brckts.surround[event.char][1] .. brckts.surround[event.char][2])
                self.state:moveCursors('left')
            end
            
            self.state:ende()
            
            return
        else 
            if valid(self.state.s.selections) then 
                if valid(self.stringDelimiterSpans) then 
                    local delrngs = belt.normalizeRanges(belt.rangesForSpans(self.stringDelimiterSpans))
                    if (delrngs == self.state.s.selections) then 
                        return 'unhandled' -- don't surround selected string delimiters
                    end
                end
                
                return self.state.surroundSelection(event.char, brckts.surround[event.char])
            end
            
            local nsegl = belt.segsForPositions(self.state.s.lines, self.state.s.cursors) -- collect all graphemes at the cursors
            local nsegs = util.uniq(nsegl) -- get set of graphemes
            
            for _, seg in ipairs(nsegs) do 
                if (brckts.surround[event.char][2] == event.char) then 
                    if (seg == event.char) then 
                        self.state:moveCursors('right') -- move cursor over existing bracket   # ϝ𝓊𝜏𝓊ɼϵ ⫙ϵ 𝖍𝕒𝜏ϵ ⟅𝚒𝛋ϵ⟅𝛾 
                        return -- to not disturb manual closing       # 𝔟𝓊𝜏 𝔭ℜϵ𝖘ϵ∩𝜏 ⫙ϵ ⟅⩜𝓏𝛾 ;)
                    end
                end
                
                if not array('', undefined, ' ', '}', ']', ')'):has(seg) then 
                    return "unhandled" -- abort the insertion
                end
            end
            
            for _, pos in ipairs(self.state.s.cursors) do 
                if belt.isUnbalancedPosition(self.state.s.lines, pos, event.char) then 
                    return "unhandled"
                end
            end
            
            self.state:insert(brckts.surround[event.char][1] .. brckts.surround[event.char][2]) -- insert empty bracket pair
            self.state:moveCursors('left') -- move cursors inside pair
            return
        end
        
        return 'unhandled'
    end

return brckts